src/main/scala/org/apache/spark/sql/TiStrategy.scala

```diff
@@ -263,129 +263,69 @@ class TiStrategy(context: SQLContext) extends Strategy with Logging {
+      deterministicAggAliases.getOrElse(e, Alias(e, e.toString())())
+    }
 
     val residualAggregateExpressions = aggregateExpressions.map { aggExpr =>
+      // As `aggExpr` is being pushing down to TiKV, we need to replace the original Catalyst
+      // aggregate expressions with new ones that merges the partial aggregation results returned by
+      // TiKV.
+      //
+      // NOTE: Unlike simple aggregate functions (e.g., `Max`, `Min`, etc.), `Count` must be
+      // replaced with a `Sum` to sum up the partial counts returned by TiKV.
+
+      // An attribute referring to the partial aggregation results returned by TiKV.
+      val partialResultRef = aliasPushedPartialResult(aggExpr).toAttribute
+
       aggExpr.aggregateFunction match {
-        // here aggExpr is the original AggregationExpression
-        // and will be pushed down to TiKV
-        case Max(_)   => newAggregate(Max(toAlias(aggExpr).toAttribute), aggExpr)
-        case Min(_)   => newAggregate(Min(toAlias(aggExpr).toAttribute), aggExpr)
-        case Count(_) => newAggregate(Sum(toAlias(aggExpr).toAttribute), aggExpr)
-        case Sum(_)   => newAggregate(Sum(toAlias(aggExpr).toAttribute), aggExpr)
-        case First(_, ignoreNullsExpr) =>
-          newAggregate(First(toAlias(aggExpr).toAttribute, ignoreNullsExpr), aggExpr)
-        case _ => aggExpr
-      }
-    } flatMap { aggExpr =>
-      aggExpr match {
-        // We have to separate average into sum and count
-        // and for outside expression such as average(x) + 1,
-        // Spark has lift agg + 1 up to resultExpressions
-        // We need to modify the reference there as well to forge
-        // Divide(sum/count) + 1
-        case aggExpr @ AggregateExpression(Average(ref), _, _, _) =>
-          // Need a type promotion
-          val sumToPush = newAggregate(Sum(ref), aggExpr)
-          val countToPush = newAggregate(Count(ref), aggExpr)
-
-          // Need a new expression id since they are not simply rewrite as above
-          val sumFinal = newAggregateWithId(Sum(toAlias(sumToPush).toAttribute), aggExpr)
-          val countFinal = newAggregateWithId(Sum(toAlias(countToPush).toAttribute), aggExpr)
-
-          avgPushdownRewriteMap(aggExpr.resultId) = List(sumToPush, countToPush)
-          avgFinalRewriteMap(aggExpr.resultId) = List(sumFinal, countFinal)
-          List(sumFinal, countFinal)
-        case _ => aggExpr :: Nil
+        case e: Max     => aggExpr.copy(aggregateFunction = e.copy(child = partialResultRef))
+        case e: Min     => aggExpr.copy(aggregateFunction = e.copy(child = partialResultRef))
+        case e: Sum     => aggExpr.copy(aggregateFunction = e.copy(child = partialResultRef))
+        case e: First   => aggExpr.copy(aggregateFunction = e.copy(child = partialResultRef))
+        case _: Count   => aggExpr.copy(aggregateFunction = Sum(partialResultRef))
+        case _: Average => throw new IllegalStateException("All AVGs should have been rewritten.")
```

Comment:
It seems we should rewrite `Average` to `sum`/`count` as the original logic rather than throw an exception here.

Such querys like:
```sql
select avg(any_col) from any_table
```
will trigger `IllegalStateException`.

Reply:
I think it's fixed now (but I cannot test it...).

Reply:
Sorry for the inconvenience, I shall do tests for you.

Reply:
On the other hand, I'm happy to code w/o writing any tests, LOL